File: json-schema/research/format_materialize_clock/runner.rb
# Probe: MATERIALIZE model (B) in Ruby via stdlib DateTime/Date. PROSPECTIVE.
# Parse each validated wire string, re-serialize via the GENERATOR-OWNED
# serializer (RFC 3339, original offset preserved with +00:00/-00:00 -> Z, T/Z
# uppercased on the parse path, fractional seconds at the value's own precision
# with trailing zeros trimmed) -- NO TRUNCATION. ruby runner.rb corpus.json
#
#   date-time -> DateTime  (offset + nanosecond preserved via Rational fraction)
#   date      -> Date      (YYYY-MM-DD, lossless)
#   time      -> UNSUPPORTED (Ruby has no time-of-day-only type; Time carries a date)
#
# FINDING: DateTime.rfc3339 SILENTLY CLAMPS leap :60 -> :59 (no error), like JS
# Temporal. The materialized grammar rejects :60 at VALIDATION; we model that
# with an explicit guard so the leap row is non-materializing, not corrupted.
require "json"
require "date"

ENGINE = "ruby"

def emit(o)
  puts JSON.generate(o)
end

def reject_leap(wire)
  raise "leap second :60 rejected by materialized grammar" if wire =~ /:60(\.\d+)?(Z|[+-]\d{2}:\d{2})?$/i
end

def frac_nanos(nanos)
  return "" if nanos.zero?
  "." + format("%09d", nanos).sub(/0+$/, "")
end

def offset_str(secs)
  return "Z" if secs.zero?
  sign = secs < 0 ? "-" : "+"
  secs = secs.abs
  format("%s%02d:%02d", sign, secs / 3600, (secs % 3600) / 60)
end

def canon_datetime(wire)
  reject_leap(wire)
  d = DateTime.rfc3339(wire.upcase) # preserves offset; requires an offset
  nanos = (d.sec_fraction * 1_000_000_000).to_i
  off = (d.offset * 86_400).to_i
  format("%04d-%02d-%02dT%02d:%02d:%02d%s%s",
         d.year, d.month, d.day, d.hour, d.min, d.sec,
         frac_nanos(nanos), offset_str(off))
end

def canon_date(wire)
  d = Date.iso8601(wire)
  format("%04d-%02d-%02d", d.year, d.month, d.day)
end

def canon_time(_wire)
  raise "UNSUPPORTED: Ruby has no time-of-day-only type"
end

def run(rows, fmt, fn)
  rows.each do |r|
    begin
      emit({ id: r["id"], engine: ENGINE, format: fmt, canonical: fn.call(r["wire"]), err: "" })
    rescue => e
      emit({ id: r["id"], engine: ENGINE, format: fmt, canonical: "", err: "#{e.class}: #{e.message}" })
    end
  end
end

c = JSON.parse(File.read(ARGV[0]))
run(c["date-time"], "date-time", method(:canon_datetime))
run(c["date"], "date", method(:canon_date))
run(c["time"], "time", method(:canon_time))

File: json-schema/research/format_materialize_clock/runner.go
// Probe: MATERIALIZE model (B). Parse each validated wire string into Go's
// stdlib typed construct, then re-serialize via the GENERATOR-OWNED serializer
// (RFC 3339, original offset preserved with +00:00/-00:00 -> Z, T/Z uppercased
// on the parse path, fractional seconds at the value's own precision with
// trailing zeros trimmed, no fractional part when zero) -- NO TRUNCATION.
//
//	date-time -> time.Time            (offset + nanosecond preserved, lossless)
//	date      -> time.Time (phantom)  (YYYY-MM-DD, lossless)
//	time      -> time.Time (phantom)  (offset preserved when present, lossless)
//
// go run runner.go corpus.json
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

type row struct {
	ID   string `json:"id"`
	Wire string `json:"wire"`
}
type corpus struct {
	DateTime []row `json:"date-time"`
	Date     []row `json:"date"`
	Time     []row `json:"time"`
}
type out struct {
	ID        string `json:"id"`
	Engine    string `json:"engine"`
	Format    string `json:"format"`
	Canonical string `json:"canonical"` // serialized bytes, or "" if cannot materialize
	Err       string `json:"err"`
}

const engine = "go"

func emit(o out) {
	b, _ := json.Marshal(o)
	fmt.Println(string(b))
}

// fractional part of a nanosecond count: ".ddd" with trailing zeros trimmed,
// or "" when zero.
func fracNanos(nanos int) string {
	if nanos == 0 {
		return ""
	}
	s := fmt.Sprintf("%09d", nanos)
	s = strings.TrimRight(s, "0")
	return "." + s
}

// offset from a count of seconds: "Z" for zero, else "+HH:MM" / "-HH:MM".
func offsetStr(secs int) string {
	if secs == 0 {
		return "Z"
	}
	sign := "+"
	if secs < 0 {
		sign = "-"
		secs = -secs
	}
	return fmt.Sprintf("%s%02d:%02d", sign, secs/3600, (secs%3600)/60)
}

func main() {
	data, _ := os.ReadFile(os.Args[1])
	var c corpus
	json.Unmarshal(data, &c)

	// date-time : time.Time. Parse path uppercases the case-insensitive t/z
	// (pinned grammar accepts lowercase; Go's native parser rejects it). Safe
	// because date-time has no other letters (offset is digits only). Go
	// REJECTS leap :60, so that row errors -> proves the narrowing.
	// time.RFC3339Nano preserves the offset (Z for zero) and nanoseconds and
	// trims trailing fractional zeros -- exactly the generator-owned form.
	for _, r := range c.DateTime {
		t, err := time.Parse(time.RFC3339Nano, strings.ToUpper(r.Wire))
		if err != nil {
			emit(out{r.ID, engine, "date-time", "", err.Error()})
			continue
		}
		emit(out{r.ID, engine, "date-time", t.Format(time.RFC3339Nano), ""})
	}

	// date : Go has no date-only type; time.Time carries a phantom time-of-day.
	for _, r := range c.Date {
		t, err := time.Parse("2006-01-02", r.Wire)
		if err != nil {
			emit(out{r.ID, engine, "date", "", err.Error()})
			continue
		}
		emit(out{r.ID, engine, "date", t.Format("2006-01-02"), ""})
	}

	// time : Go has no time-of-day type; time.Time carries a phantom date. The
	// offset is PRESERVED (rides in the zone) when present; an offset-less value
	// stays offset-less. Serialize manually so trailing fractional zeros are
	// trimmed and +00:00/-00:00 -> Z consistently.
	for _, r := range c.Time {
		w := strings.ToUpper(r.Wire)
		var t time.Time
		var err error
		hasOffset := true
		t, err = time.Parse("15:04:05.999999999Z07:00", w)
		if err != nil {
			hasOffset = false
			t, err = time.Parse("15:04:05.999999999", w)
		}
		if err != nil {
			emit(out{r.ID, engine, "time", "", err.Error()})
			continue
		}
		s := fmt.Sprintf("%02d:%02d:%02d%s", t.Hour(), t.Minute(), t.Second(), fracNanos(t.Nanosecond()))
		if hasOffset {
			_, off := t.Zone()
			s += offsetStr(off)
		}
		emit(out{r.ID, engine, "time", s, ""})
	}
}
